### DISCUSSION TITLE
online vs. asleep

### DISCUSSION THREAD

**Discussion by emufan** [alandtse/tesla#387] [Q&A]
Can someone explain, why there is since one of the last updates **asleep** in addition to **online**, it is reporting only the other binary state?

![image](https://user-images.githubusercontent.com/22775515/205014758-6b3aa159-c03e-421e-83cc-8b76b6aace9a.png)

And I think that online was some time ago before "server online", HA-serverconnecction and not only notasleep as now and shown above. Then I think it makes sense to have both, because different information given: car asleep and server-ha-connection online.

**Comment by carleeno** [comment]
Hi, it's possible the vehicle is offline but not asleep, the api reports the vehicle state as `online`, `asleep`, or `offline` (or `unavailable`, I can't remember at this moment). The Online sensor is state == 'online', and the Asleep sensor is state == 'asleep'

You are correct, there are times when the car is offline due to a connection issue, but not asleep. The connection issue could be server-side but it could also be that the vehicle has no wifi or LTE signal.

**Reply by emufan** [reply]
> it's possible the vehicle is offline but not asleep

But this wouldn't be reported anyways. Or how should this work? If it is offline, it would not be possible to report the asleep state, so this would only correct by chance - or not. But no guarantee to be true.

So as long there is no case, where is asleep, but online, for me HA will always report the same, but invers state, which seems to be somehow redundant.

BTW is the API getting the state directly from the car or from a server (which gets the state from the car)?

**Reply by carleeno** [reply]
The Tesla servers keep track of whether it's asleep vs unavailable. When asleep, the car is still connected as you can wake it up remotely. In my case because I don't have good connectivity to my car when it's parked at home, it will sometimes report that it's offline but not asleep...so it's not always the case that asleep and online are perfectly inverted.

We added the new asleep sensor to provide ability to create automations for the different scenarios.

Maybe the confusion is that the online sensor should remain online when it's asleep, but this behavior change would break existing automations unfortunately.

**Reply by emufan** [reply]
> it will sometimes report that it's offline but not asleep

But you cannot rely on this, If it is offline/not connected but in the meantime asleep or not asleep, the server will not know.

> Maybe the confusion is that the online sensor should remain online when it's asleep, but this behavior change would break existing automations unfortunately.

I'm somehow sure, taht it was this way. Here it was online (and updated) all the time until one of the last updated. So I always thought, that it is the server-connection or if the car is connected to the server. It started to go offline only now and then inverted to asleep.

> When asleep, the car is still connected as you can wake it up remotely.

Yes. But then it should be online/connected as in the last example and not offline aka disconnected. For me offline is, when no connection is possible.

**Reply by carleeno** [reply]
>  If it is offline/not connected but in the meantime asleep or not asleep, the server will not know.

True, it assumes the last state, just like all the other sensors do.

> I'm somehow sure, taht it was this way. Here it was online (and updated) all the time until one of the last updated. So I always thought, that it is the server-connection or if the car is connected to the server. It started to go offline only now and then inverted to asleep.

You may be thinking of v2, I can't remember what it did back then. But in v 3.0 and 3.1, there was only the `online` sensor, and it was false when the car was asleep OR offline, which meant that we couldn't tell the difference between the two.

> Yes. But then it should be online/connected as in the last example and not offline aka disconnected. For me offline is, when no connection is possible.

We both agree on this, `online` should probably stay true when it is asleep, and only turn false when the tesla servers say the car is truly offline/unavailable. If you'd like this change, can you create an issue that the online sensor should stay true when asleep since it is still technically online (and wake-able)? I can work on the change if so and see if we can get approval, although it might break some automations which is unfortunate.

**Reply by emufan** [reply]
Did so.

https://github.com/alandtse/tesla/issues/389
